docs/source/kinds_of_types.rst

```diff
@@ -721,18 +721,31 @@ For more details, see :ref:`type-variable-value-restriction`.
 Generators
 **********
 
-A basic generator that only yields values can be annotated as having a return
+A basic generator that only yields values can be succinctly annotated as having a return
 type of either :py:class:`Iterator[YieldType] <typing.Iterator>` or :py:class:`Iterable[YieldType] <typing.Iterable>`. For example:
 
 .. code-block:: python
 
    def squares(n: int) -> Iterator[int]:
        for i in range(n):
            yield i * i
+           
+In general, :py:class:`Iterator[YieldType] <typing.Iterator>` should be preferred over
+:py:class:`Iterable[YieldType] <typing.Iterable>` for the return type of a generator function,
+as it is more precise regarding the type of the returned object. All iterator objects have a
+:py:meth:`~iterator.__next__` method, so using ``Iterator`` as the return type alerts mypy to
+the fact that :py:func:`next` can be called on the object returned from the generator function.
+Using ``Iterable`` runs the risk of mypy raising false-positive errors if a user attempts to call
+`next()` on the returned object.
+
+However, in some cases, it may be considered an implementation detail that ``next()`` can be
+called on the object returned from the generator function. In these cases, it may be preferable
+to use ``Iterable`` instead of ``Iterator``.
```

Comment:
Nit rewording. My goals are a) trying to keep things brief and the recommendation clear, since new users' eyes tend to glaze over, b) calling out the rule of thumb explicitly
```suggestion
A good rule of thumb is to annotate functions with the most specific return type
that doesn't leak incidental implementation detail. In keeping with this, prefer
:py:class:`Iterator[YieldType] <typing.Iterator>` over
:py:class:`Iterable[YieldType] <typing.Iterable>`. In particular, using
``Iterator`` lets mypy know users are allowed to call :py:func:`next` on the
object returned from the generator function.
```

Reply:
My wording was too verbose, but I found "A good rule of thumb is to annotate functions with the most specific return type
that doesn't leak incidental implementation detail" to be too dense. What do you think of the latest iteration? 🙂